File: src/codex_subagent_kit/release_versioning.py
from __future__ import annotations

import re
from typing import Iterable


SEMVER_PATTERN = re.compile(r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$")
BREAKING_SUBJECT_PATTERN = re.compile(r"^[a-zA-Z]+(?:\([^)]+\))?!:")
FEATURE_SUBJECT_PATTERN = re.compile(r"^feat(?:\([^)]+\))?:")


def parse_semver(value: str) -> tuple[int, int, int]:
    match = SEMVER_PATTERN.fullmatch(value.strip())
    if not match:
        raise ValueError(f"invalid semantic version: {value}")
    return tuple(int(part) for part in match.groups())


def bump_semver(version: str, bump: str) -> str:
    major, minor, patch = parse_semver(version)
    if bump == "major":
        return f"{major + 1}.0.0"
    if bump == "minor":
        return f"{major}.{minor + 1}.0"
    if bump == "patch":
        return f"{major}.{minor}.{patch + 1}"
    raise ValueError(f"unsupported bump kind: {bump}")


def classify_bump(commit_messages: Iterable[str]) -> str:
    normalized = [message.strip() for message in commit_messages if message.strip()]
    if not normalized:
        return "patch"

    for message in normalized:
        subject = message.splitlines()[0].strip()
        if "BREAKING CHANGE:" in message or BREAKING_SUBJECT_PATTERN.match(subject):
            return "major"

    for message in normalized:
        subject = message.splitlines()[0].strip()
        if FEATURE_SUBJECT_PATTERN.match(subject):
            return "minor"

    return "patch"


def compute_next_version(base_version: str, commit_messages: Iterable[str]) -> str:
    return bump_semver(base_version, classify_bump(commit_messages))

File: packages/codex-subagent-kit/src/release-versioning.ts
const SEMVER_PATTERN = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$/;
const RELEASE_LABELS = ["release:major", "release:minor", "release:patch", "release:none"] as const;

export type SemverBump = "major" | "minor" | "patch";
export type ReleaseDecision = SemverBump | "none";

export function parseSemver(value: string): [number, number, number] {
  const match = value.trim().match(SEMVER_PATTERN);
  if (!match) {
    throw new Error(`invalid semantic version: ${value}`);
  }
  return [Number(match[1]), Number(match[2]), Number(match[3])];
}

export function bumpSemver(version: string, bump: SemverBump): string {
  const [major, minor, patch] = parseSemver(version);
  if (bump === "major") {
    return `${major + 1}.0.0`;
  }
  if (bump === "minor") {
    return `${major}.${minor + 1}.0`;
  }
  return `${major}.${minor}.${patch + 1}`;
}

export function classifyReleaseLabels(labels: Iterable<string>): ReleaseDecision {
  const normalized = [...labels]
    .map((label) => label.trim())
    .filter(Boolean);
  const recognized = [...new Set(normalized.filter((label): label is (typeof RELEASE_LABELS)[number] =>
    (RELEASE_LABELS as readonly string[]).includes(label),
  ))];

  if (recognized.length === 0) {
    return "none";
  }
  if (recognized.length > 1) {
    throw new Error(`conflicting release labels: ${recognized.join(", ")}`);
  }

  const [label] = recognized;
  if (label === "release:none") {
    return "none";
  }
  if (label === "release:major") {
    return "major";
  }
  if (label === "release:minor") {
    return "minor";
  }
  return "patch";
}

export function computeNextVersionFromLabels(
  baseVersion: string,
  labels: Iterable<string>,
  options?: { initialRelease?: boolean },
): string | null {
  const decision = classifyReleaseLabels(labels);
  if (decision === "none") {
    return null;
  }
  if (options?.initialRelease) {
    return baseVersion;
  }
  return bumpSemver(baseVersion, decision);
}
